File: dfsmatrix/prog.swift
import Foundation;
let n = Int(readLine()!)!
let m = Int(readLine()!)!

class Node {
  var id: String
  var address: (x: Int, y: Int)
  var value: Int
  var graph_id: String?

  init(id: String, address: (Int, Int), value: Int) {
    self.id = id
    self.address = address
    self.value = value
  }
}

var nodes = [String: Node]()

func graph_size(size: inout Int, node: Node, nodes: inout [String: Node]) -> Int {  
  for x in -1..<2 {
    for y in -1..<2 {
      let neighbor_id = "\(node.address.x + x)\(node.address.y + y)"
      if nodes[neighbor_id] != nil, let neighbor_node = nodes[neighbor_id] {
        if(neighbor_node.value == 0 || neighbor_node.graph_id != nil){ continue }
        neighbor_node.graph_id = node.graph_id
        size+=1
        size = graph_size(size: &size, node: neighbor_node, nodes: &nodes)
      }
    }
  }
  return size
}

for x in 0..<n {
  var row = readLine()!.components(separatedBy: [" "])
  for y in 0..<row.count {
    var id = "\(x)\(y)"
    nodes[id] = Node(id: id, address: (x,y), value: Int(row[y])!)
  }
}

var largest = 0
for (id,node) in nodes {
  if(node.value == 0 || node.graph_id != nil) { continue }
  node.graph_id = node.id
  var size = 1
  size = graph_size(size: &size, node: node, nodes: &nodes)
  largest = largest > size ? largest : size
}

print("\(largest)")

File: dfsmatrix/prog.rb
#!/bin/ruby

n = gets.strip.to_i
m = gets.strip.to_i

nodes = {}
n.times do |i| 
  gets.strip.split(" ").each_with_index.map do |value,k| 
    nodes[:"#{i}#{k}"] = { address: :"#{i}#{k}", value: value.to_i } 
  end
end

def graph_size(nodes,node,size)
  neighbor_nodes = []
  node_address_p = node[:address].to_s.chars.map(&:to_i)
  (-1..1).each { |i| (-1..1).each { |k| neighbor_nodes << nodes[[node_address_p[0] + i, node_address_p[1] + k].join.to_s.to_sym] } }
  neighbor_nodes.each do |neighbor_node|
    next if !neighbor_node || neighbor_node[:value] == 0 || !!neighbor_node[:graph_id]
    neighbor_node[:graph_id] = node[:graph_id]
    size = graph_size(nodes,neighbor_node,size+1)
  end
  return size
end

largest = 0
nodes.each do |address,node|
  next if node[:value] == 0 || !!node[:graph_id]
  node[:graph_id] = node[:address]
  gs = graph_size(nodes,node,1)
  largest = largest > gs ? largest : gs
end

puts largest